[Style] Convert the 'will-change' property to strong style types

weinig opened the pull request:
#### c84169878e1555ca501a0be62d1ee6959491013b
<pre>
[Style] Convert the &apos;will-change&apos; property to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=301433">https://bugs.webkit.org/show_bug.cgi?id=301433</a>

Reviewed by Darin Adler.

Converts the &apos;will-change&apos; property to use strong style types.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderElementInlines.h:
* Source/WebCore/rendering/RenderElementStyleInlines.h:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderInline.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/WillChangeData.cpp: Removed.
* Source/WebCore/rendering/style/WillChangeData.h: Removed.
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/will-change/StyleWillChange.cpp: Added.
* Source/WebCore/style/values/will-change/StyleWillChange.h: Added.

Canonical link: <a href="https://commits.webkit.org/302389@main">https://commits.webkit.org/302389@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/069f4f003cd65025cf76deb7e1ad4041d8da202d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128934 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39765 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136313 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80291 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130805 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1127 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1066 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98157 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66066 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115495 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78785 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/790 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33609 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79593 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109227 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34105 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138779 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/990 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/964 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106693 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 43 flakes 91 failures; Uploaded test results; 26 flakes 64 failures; Compiled WebKit; Running layout-tests-repeat-failures-without-change") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111834 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106508 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27120 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/819 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30355 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53451 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1064 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64384 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/901 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/957 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/996 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1908,10 +1908,8 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle, WTF::DefaultedOp
-
-    bool willChangeCreatesStackingContext() const;
+    inline const Style::WillChange& willChange() const;
+    void setWillChange(Style::WillChange&&);
```
```suggestion
    inline void setWillChange(Style::WillChange&&);
```

darinadler commented on Source/WebCore/style/values/will-change/StyleWillChange.cpp:
```diff
@@ -0,0 +1,245 @@
+bool WillChangeAnimatableFeature::propertyCreatesStackingContext(CSSPropertyID property)
+{
+    switch (property) {
+    case CSSPropertyPerspective:
+    case CSSPropertyWebkitPerspective:
+    case CSSPropertyScale:
+    case CSSPropertyRotate:
+    case CSSPropertyTranslate:
+    case CSSPropertyTransform:
+    case CSSPropertyTransformStyle:
+    case CSSPropertyOffsetPath:
+    case CSSPropertyClipPath:
+    case CSSPropertyMask:
+    case CSSPropertyWebkitMask:
+    case CSSPropertyOpacity:
+    case CSSPropertyPosition:
+    case CSSPropertyZIndex:
+    case CSSPropertyWebkitBoxReflect:
+    case CSSPropertyMixBlendMode:
+    case CSSPropertyIsolation:
+    case CSSPropertyFilter:
+    case CSSPropertyBackdropFilter:
+    case CSSPropertyWebkitBackdropFilter:
+    case CSSPropertyMaskImage:
+    case CSSPropertyMaskBorder:
+    case CSSPropertyWebkitMaskBoxImage:
+#if ENABLE(WEBKIT_OVERFLOW_SCROLLING_CSS_PROPERTY)
+    case CSSPropertyWebkitOverflowScrolling:
+#endif
+    case CSSPropertyViewTransitionName:
+    case CSSPropertyContain:
```
What's our strategy to know whether this list is complete? Was this just moved from elsewhere? Why isn't the list sorted alphabetically since it's so long?

darinadler commented on Source/WebCore/style/values/will-change/StyleWillChange.cpp:
```diff
@@ -0,0 +1,245 @@
+
+bool WillChangeAnimatableFeatures::Data::createsContainingBlockForOutOfFlowPositioned(bool isRootElement) const
+{
+    return containsProperty(CSSPropertyPerspective)
+        // CSS transforms
+        || containsProperty(CSSPropertyTransform)
+        || containsProperty(CSSPropertyTransformStyle)
+        || containsProperty(CSSPropertyTranslate)
+        || containsProperty(CSSPropertyRotate)
+        || containsProperty(CSSPropertyScale)
+        || containsProperty(CSSPropertyOffsetPath)
+        // CSS containment
+        || containsProperty(CSSPropertyContain)
+        // CSS filter & backdrop-filter
+        || (containsProperty(CSSPropertyBackdropFilter) && !isRootElement)
+        || (containsProperty(CSSPropertyWebkitBackdropFilter) && !isRootElement)
+        || (containsProperty(CSSPropertyFilter) && !isRootElement);
```
Is this OK for performance? Iterating 11 times seems possibly not good?

weinig replied:
Just moved from WillChangeData. A bit set could definitely help here, though unclear if it’s necessary.

weinig committed:
[Style] Convert the 'will-change' property to strong style types
https://bugs.webkit.org/show_bug.cgi?id=301433

Reviewed by Darin Adler.

Converts the 'will-change' property to use strong style types.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderElementInlines.h:
* Source/WebCore/rendering/RenderElementStyleInlines.h:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderInline.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/WillChangeData.cpp: Removed.
* Source/WebCore/rendering/style/WillChangeData.h: Removed.
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/will-change/StyleWillChange.cpp: Added.
* Source/WebCore/style/values/will-change/StyleWillChange.h: Added.

Canonical link: https://commits.webkit.org/302389@main
